### ISSUE TITLE
StarcounterEnvironment.AppName is not set inside .NET threads

### ISSUE THREAD

**Issue by alemoi** [Starcounter/Starcounter#3149]
As we know `StarcounterEnvironment.AppName` is crucial, many concepts rely on correct app name. The following scenario is not working now:

1. Start a .NET thread inside main() -> inside this new thread "StarcounterEnvironment.AppName" is always null, which is incorrect.
2. Inside this .NET thread do dbSession.RunAsync/Sync ->  "StarcounterEnvironment.AppName" will be null even inside the scheduler, which is incorrect.

I created a special test that covers these situations: TestAppName.cs in SimpleIndependentTests solution in Level1. @per-samuelsson let's have a discussion on how we can solve this.

**Comment by per-samuelsson** [comment]
> @per-samuelsson let's have a discussion on how we can solve this.

I'll gladly help if I can, but I really have no idea how to do this, and what you have in mind. 

> Start a .NET thread inside main() -> inside this new thread "StarcounterEnvironment.AppName" is always null, which is incorrect.

I mean, this is very basic threading in .NET. Thread statics are / thread and of course its null by default if you have a reference. How could we ever get it not to be that? 

I guess it must somehow be set when the thread attach to a scheduler? I dont know the full picture of how this thread-local app name is assigned really, and if all schedulers always operate in one, etc. But thats the best idea I can come up with.

For me, extra threads like this is not something we want to promote normally. And based on that, we could enforce the user to specify where to get the app name from I guess. Don't think it should be more to it than that. We dont want to get to fancy.

**Comment by per-samuelsson** [comment]
> For me, extra threads like this is not something we want to promote normally. And based on that, we could enforce the user to specify where to get the app name from I guess. Don't think it should be more to it than that. We dont want to get to fancy.

Something like this.

```C#
class Db {
    public static Thread StartThread(Action a) {
        var t = new Thread((name) => {
            StarcounterEnvironment.AppName = (string)name;
            a();
        });

        t.Start(StarcounterEnvironment.AppName);
        return t;
    }
}
```

And used like this:

```C#
Db.StartThread(() => { ... });
```

And if users still do `new Thread(...)`, they are on their own. 

Maybe, in some future version, we could enforce it by means of the weaver. But lets try postpone that until we get to the brand new Roslyn world.

**Comment by bigwad** [comment]
If that matters, I used to avoid this problem using StarcounterEnvironment.RunWithinApplication.
For example:
    public async Task Run(byte scheduler)
    {
      string app_name = Starcounter.Internal.StarcounterEnvironment.AppName;
      await Task.Yield();

      new DbSession().RunSync(() =>
      {
        Starcounter.Internal.StarcounterEnvironment.RunWithinApplication(app_name,
          () =>
          {
             //some db-related work
          });
      }, scheduler);
    }

**Comment by alemoi** [comment]
Yes, if we can provide some wrapping API around .NET threads (including simpler start delegate) then its fine. I was just thinking that we can not restrict users for that, because I thought there can be many ways to run things outside Starcounter. @dan31 what do you think about limiting user API running outside Starcounter scheduler?

**Comment by bigwad** [comment]
I mean we already has the API to solve this problem and it doesn't limit user in any way. But maybe I just don't understand the topic. Then please ignore my post :)
